das/dastree/dastree.go

```diff
@@ -0,0 +1,129 @@
+
+import (
+	"fmt"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/offchainlabs/nitro/util/arbmath"
+)
+
+const BinSize = 64 * 1024 // 64 kB
+
+type bytes32 = common.Hash
+
+type node struct {
+	hash bytes32
+	size uint32
+}
+
+func RecordHash(record func(bytes32, []byte), preimage ...[]byte) bytes32 {
+	// Algorithm
+	//  1. split the preimage into 64kB bins and hash them to produce the tree's leaves
+	//  2. repeatedly hash pairs and their combined length, bubbling up any odd-one's out, to form the root
+	//
+	//            r         <=>  hash(hash(0, 1), 2, len(0:2))        step 2
+	//           / \
+	//          *   2       <=>  hash(0, 1, len(0:1)), 2              step 1
+	//         / \
+	//        0   1         <=>  0, 1, 2                              step 0
+	//
+	//  Intermediate hashes like '*' from above may be recorded via the `record` closure
+	//
+
+	unrolled := []byte{}
+	for _, slice := range preimage {
+		unrolled = append(unrolled, slice...)
+	}
+	if len(unrolled) == 0 {
+		innerKeccak := crypto.Keccak256Hash([]byte{})
+		outerKeccak := crypto.Keccak256Hash(innerKeccak.Bytes())
+		record(outerKeccak, innerKeccak.Bytes())
+		record(innerKeccak, []byte{})
+		return outerKeccak
+	}
+
+	length := uint32(len(unrolled))
+	leaves := []node{}
+	for bin := uint32(0); bin < length; bin += BinSize {
+		end := arbmath.MinUint32(bin+BinSize, length)
+		content := unrolled[bin:end]
+		innerKeccak := crypto.Keccak256Hash(content)
+		outerKeccak := crypto.Keccak256Hash(innerKeccak.Bytes())
+		record(outerKeccak, innerKeccak.Bytes())
+		record(innerKeccak, content)
+		leaves = append(leaves, node{outerKeccak, end - bin})
+	}
+
+	layer := leaves
+	for len(layer) > 1 {
+		prior := len(layer)
+		after := prior/2 + prior%2
+		paired := make([]node, after)
+		for i := 0; i < prior-1; i += 2 {
+			firstHash := layer[i].hash.Bytes()
+			otherHash := layer[i+1].hash.Bytes()
+			sizeUnder := layer[i].size + layer[i+1].size
+			dataUnder := firstHash
+			dataUnder = append(dataUnder, otherHash...)
+			dataUnder = append(dataUnder, arbmath.Uint32ToBytes(sizeUnder)...)
+			parent := node{
+				crypto.Keccak256Hash(dataUnder),
+				sizeUnder,
+			}
+			record(parent.hash, dataUnder)
+			paired[i/2] = parent
+		}
+		if prior%2 == 1 {
+			paired[after-1] = layer[prior-1]
+		}
+		layer = paired
+	}
+	return layer[0].hash
+}
+
+func Hash(preimage ...[]byte) bytes32 {
+	// Merkelizes without recording anything. All but the replay binary's DAS will call this
+	return RecordHash(func(bytes32, []byte) {}, preimage...)
+}
+
+func HashBytes(preimage ...[]byte) []byte {
+	return Hash(preimage...).Bytes()
+}
+
+func FlatHashToTreeHash(flat bytes32) bytes32 {
+	// Forms a degenerate dastree that's just a single leaf
+	// note: the inner preimage may be arbitrarily larger than the 64 kB standard
+	return crypto.Keccak256Hash(flat[:])
+}
+
+func Content(root bytes32, oracle func(bytes32) []byte) ([]byte, error) {
```

Comment:
Add a comment explaining why it's OK that this doesn't check the value of the size field, also why it doesn't check consistency of size field between parent and children.  

(Also, wondering why the size field is included, if it's not checked here.)

Reply:
The size field is for future proofing in case we'd ever want an API that allowed for looking up, say, an arbitrary byte from the preimage without exploring the entire tree. Right now there's no direct need to check sizes or canonicalization because if the committee signs an evil certificate with a non-canonical structure or malicious size data then the AnyTrust assumption has been violated and we have bigger issues

Nevertheless I've implemented the size checks and added some commentary. There's value in testing that the structure is correctly formed with consistent size data and it adds clarity to the design :)